store/src/server/api.rs

```diff
@@ -1,27 +1,49 @@
 use crate::config::StoreConfig;
 use crate::db::Db;
 use anyhow::Result;
-use miden_crypto::{merkle::TieredSmt, Felt, FieldElement};
-use miden_node_proto::store;
-use std::net::ToSocketAddrs;
-use tokio::time::Instant;
+use miden_crypto::{merkle::TieredSmt, Felt, FieldElement, hash::rpo::RpoDigest};
+use miden_node_proto::store::{self, CheckNullifiersResponse};
+use std::{net::ToSocketAddrs, sync::Arc};
+use tokio::{time::Instant, sync::{mpsc, RwLock}};
+use tokio_stream::{StreamExt, wrappers::ReceiverStream};
 use tonic::{transport::Server, Response, Status, Streaming};
 use tracing::{info, instrument};
 
 pub struct StoreApi {
     db: Db,
-    nullifier_tree: TieredSmt,
+    nullifier_tree: Arc<RwLock<TieredSmt>>,
 }
 
 #[tonic::async_trait]
 impl store::api_server::Api for StoreApi {
-    type CheckNullifiersStream = Streaming<store::CheckNullifiersResponse>;
+    type CheckNullifiersStream = ReceiverStream<Result<store::CheckNullifiersResponse, Status>>;
 
     async fn check_nullifiers(
         &self,
-        _request: tonic::Request<Streaming<store::CheckNullifiersRequest>>,
+        request: tonic::Request<Streaming<store::CheckNullifiersRequest>>,
     ) -> Result<Response<Self::CheckNullifiersStream>, Status> {
-        todo!()
+        let mut messages = request.into_inner();
+
+        let (tx, rx) = mpsc::channel(128);
+        let tree = self.nullifier_tree.clone();
+
+        tokio::spawn(async move {
+            while let Some(msg) = messages.next().await {
+                let result = if let Some(nullifier) = msg?.nullifier {
+                    let digest = RpoDigest::new([Felt::new(nullifier.d0), Felt::new(nullifier.d1), Felt::new(nullifier.d2), Felt::new(nullifier.d3)]);
+                    let leaf = tree.read().await.get_value(digest);
```

Comment:
Note: A single leaf is read at the time, meaning a concurrent write may happen. I think this is the correct behavior here, the alternative would be to copy the tree, or to forbid writes while there is a stream on-going.